Lib/contextlib.py

```diff
@@ -79,11 +82,32 @@ def _recreate_cm(self):
         return self
 
     def __call__(self, func):
-        @wraps(func)
         def inner(*args, **kwds):
             with self._recreate_cm():
                 return func(*args, **kwds)
-        return inner
+
+        def gen_inner(*args, **kwds):
+            with self._recreate_cm():
+                yield from func(*args, **kwds)
```

Comment:
```suggestion
            with self._recreate_cm(), closing(func(*args, **kwds)) as gen:
                yield from gen
```

I think we should be using `with closing()` and `async with aclosing()` in these wrappers, to ensure that any resources held by the wrapped function are cleaned up promptly.  (see e.g. [PEP-533 motivation](https://peps.python.org/pep-0533/#background-and-motivation))

Reply:
Done.